src/Halogen/Aff/Util.purs

```diff
@@ -7,33 +7,31 @@ module Halogen.Aff.Util
 
 import Prelude
 
-import Control.Monad.Aff (Aff, makeAff, runAff)
+import Control.Monad.Aff (Aff, Canceler(..), makeAff, runAff_)
 import Control.Monad.Eff (kind Effect, Eff)
 import Control.Monad.Eff.Class (liftEff)
 import Control.Monad.Eff.Exception (throwException, error)
 import Control.Monad.Error.Class (throwError)
 import Control.Monad.Except (runExcept)
-
-import Data.Maybe (Maybe(..), maybe)
-import Data.Either (either)
-import Data.Foreign (toForeign)
-
 import DOM (DOM)
-import DOM.Event.EventTarget (eventListener, addEventListener)
-import DOM.HTML.Event.EventTypes (load)
+import DOM.Event.EventTarget (addEventListener, eventListener, removeEventListener)
 import DOM.HTML (window)
+import DOM.HTML.Event.EventTypes (load)
 import DOM.HTML.Types (HTMLElement, windowToEventTarget, htmlDocumentToParentNode, readHTMLElement)
 import DOM.HTML.Window (document)
 import DOM.Node.ParentNode (QuerySelector(..), querySelector)
-
+import Data.Either (Either(..), either)
+import Data.Foreign (toForeign)
+import Data.Maybe (Maybe(..), maybe)
 import Halogen.Aff.Effects (HalogenEffects)
 
 -- | Waits for the document to load.
 awaitLoad :: forall eff. Aff (dom :: DOM | eff) Unit
-awaitLoad = makeAff \_ callback -> liftEff $
-  window
-    >>= windowToEventTarget
-    >>> addEventListener load (eventListener (\_ -> callback unit)) false
+awaitLoad = makeAff \callback -> liftEff do
+  et <- windowToEventTarget <$> window
+  let listener = eventListener (\_ -> callback (Right unit))
+  addEventListener load listener false et
+  pure $ Canceler \_ -> liftEff (removeEventListener load listener false et)
 
 -- | Waits for the document to load and then finds the `body` element.
 awaitBody :: forall eff. Aff (dom :: DOM | eff) HTMLElement
```

Comment:
We should fix `awaitBody` so that it doesn't block forever if the document has already loaded.

Reply:
Ah yeah 👍

Reply:
I took a stab at this here: https://github.com/garyb/purescript-halogen/pull/5

Reply:
What's holding this back?